### DISCUSSION TITLE
Is it possible to publish a stream that includes both video and data?

### DISCUSSION THREAD

**Discussion by joskaaaa** [bluenviron/mediamtx#357] [Questions]
I have a .TS file which contains both video and a metadata bitstream. 

For example, the ffprobe output for my file looks like this:

```
    Stream #0:0[0x100]: Video: h264 (High) ([27][0][0][0] / 0x001B), yuv420p, 1280x720, 30 tbr, 90k tbn, 180k tbc
    Stream #0:1[0x101]: Data: klv (KLVA / 0x41564C4B)
```

I would like to publish this stream - but attempting to do so with the basic copy command (even with the copy_unknown flag) leaves me with:

```
Output #0, rtsp, to 'rtsp://192.168.1.71:8554/mystream':
  Metadata:
    encoder         : Lavf57.38.102
    Stream #0:0: Video: h264 ([27][0][0][0] / 0x001B), yuv420p, 1280x720, q=2-31, 30 tbr, 90k tbn, 90k tbc
    Stream #0:1: Data: klv (KLVA / 0x41564C4B)
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #0:1 -> #0:1 (copy)
Could not write header for output file #0 (incorrect codec parameters ?): Server returned 400 Bad Request
```

Is there any way to coerce rtsp-simple-server to publish a data stream?

**Comment by aler9** [comment]
Hello, the problem is that the server must know the clock rate of each track (it's the only requirement), and your track probably doesn't provide any. Anyway, if you provide a log of the server with the `logLevel: debug` parameter, we might be able to find the problem and support these kind of tracks too.

**Comment by gediminasgu** [comment]
Here is an example of how to reproduce it:
```
ffmpeg -re -stream_loop -1 -i "Night Flight IR.mpg" -map: 0:0 -map 0:1 -c copy -f rtsp rtsp://127.0.0.1:8554/mystream
```
The source file could be downloaded from here: https://samples.ffmpeg.org/MPEG2/mpegts-klv/Night%20Flight%20IR.mpg

Here is the server log:
```
2021/06/24 10:39:36 I [0/0] [RTSP] [conn 127.0.0.1:57670] opened
2021/06/24 10:39:36 D [0/0] [RTSP] [conn 127.0.0.1:57670] [c->s] OPTIONS rtsp://127.0.0.1:8554/mystream RTSP/1.0
CSeq: 1
User-Agent: Lavf58.26.100


2021/06/24 10:39:36 D [0/0] [RTSP] [conn 127.0.0.1:57670] [s->c] RTSP/1.0 200 OK
CSeq: 1
Public: DESCRIBE, ANNOUNCE, SETUP, PLAY, RECORD, PAUSE, GET_PARAMETER, TEARDOWN
Server: gortsplib


2021/06/24 10:39:39 D [0/0] [RTSP] [conn 127.0.0.1:57670] [c->s] ANNOUNCE rtsp://127.0.0.1:8554/mystream RTSP/1.0
CSeq: 2
Content-Length: 326
Content-Type: application/sdp
User-Agent: Lavf58.26.100

v=0
o=- 0 0 IN IP4 127.0.0.1
s=No Name
c=IN IP4 127.0.0.1
t=0 0
a=tool:libavformat 58.26.100
m=video 0 RTP/AVP 96
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; sprop-parameter-sets=Z00AKZp0AoAtyA==,aO4CWIAA; profile-level-id=4D0029
a=control:streamid=0
m=application 0 RTP/AVP 97
a=control:streamid=1

2021/06/24 10:39:39 I [0/0] [RTSP] [session 2926217394] opened by 127.0.0.1:57670
2021/06/24 10:39:39 D [0/0] [RTSP] [conn 127.0.0.1:57670] [s->c] RTSP/1.0 400 Bad Request
CSeq: 2
Server: gortsplib


2021/06/24 10:39:39 I [0/0] [RTSP] [conn 127.0.0.1:57670] ERR: invalid SDP: unable to get clock rate of track 1: attribute 'rtpmap' not found
2021/06/24 10:39:39 I [0/0] [RTSP] [conn 127.0.0.1:57670] closed
2021/06/24 10:39:39 I [0/0] [RTSP] [session 2926217394] closed
```

**Reply by aler9** [reply]
Hello, i removed the clock rate check from the server, but ffmpeg crashes anyway with this message:

```
[rtp @ 0x5571f9600d00] Unsupported codec klv
Could not write header for output file #0 (incorrect codec parameters ?): Operation not permitted
```

Until ffmpeg is able to mux KLV into RTP, nothing can be done.

<s>Gstreamer is able to mux KLV into RTP, but it's not able to demux KLV from mpeg, so here we go again...</s>

**Reply by gediminasgu** [reply]
Thank you for your answer. Yes, actually Gstreamer worked here and I can get a video with KLV on the other end. For those who are interested below is an example of how to send video with KLV:
```
gst-launch-1.0 filesrc location="Night Flight IR.mpg" ! tsparse set-timestamps=true ! rtspclientsink location=rtsp://127.0.0.1:8554/mystream
```

**Reply by lars18th** [reply]
Hi @gediminasgu ,

The cause because you can do with gstreamer is because it sends the streams encapsulated in a MPEG-TS. Then the server only receives one track. The gstreamer has support to send MPEG-TS over RTSP, but ffmpeg lacks (at time) this functionality. I've post one request to improve the MPEG-TS support in rtsp-simple-server. However, I'm also interested in suport KLV streams (in general data streams, so not audio and video) over demuxed (plain) RTSP sessions.

Regards.

**Comment by lars18th** [comment]
Hi @aler9 ,

I'm working on solving this trouble. And the root problem to garantee the interoperativility is what is doing the rtsp-simple-server to handle raw streams with timestamps. As I discovered the SDP with this type of stream will contains this particular line: `a=rtpmap:96 raw/90000`. So the question is if you support the "raw" type at 90000Hz ? If not, then please could you add support for it? I'll try to provide a client that will announce such data.

Thank you.

**Reply by aler9** [reply]
>  you support the "raw" type at 90000Hz ?

yes, we can add support for it

**Comment by av-marc** [comment]
Does anyone know whether the issue of ffmpeg not being able to mux KLV into RTP has been fixed? I am not using ffmpeg but I am using libav to generate an rtsp stream with both video and klv data in it. But to no avail so far. It is able to mux it into an output.ts file, but rtsp does not work. Same error: 

Client:
Could not write header for output file #0 (incorrect codec parameters ?): Server returned 400 Bad Request

Server:

invalid SDP: unable to get clock rate of track 1: attribute 'rtpmap' not found

**Reply by aler9** [reply]
In my opinion you can either contact FFmpeg authors to get this feature developed or you can use [gortsplib](https://github.com/bluenviron/gortsplib) to connect to the server through RTSP and stream video and KLV on two separate tracks and with any RTP encoding you want.

[a comment by an automated account is not shown]
